test-e2e/test_zookeeper_backup.py

```diff
@@ -0,0 +1,278 @@
+            'dcos-zk',
+            'backup',
+            str(backup_remote_path),
+            '-v',
+        ],
+        output=Output.LOG_AND_CAPTURE,
+    )
+
+    master.run(args=['systemctl', 'start', 'dcos-exhibitor'])
+
+    master.download_file(
+        remote_path=backup_remote_path,
+        local_path=backup_local_path,
+    )
+
+    master.run(args=['rm', str(backup_remote_path)])
+
+
+def _do_restore(all_masters: Set[Node], backup_local_path: Path) -> None:
+    """
+    Automated ZooKeeper restore from backup procedure.
+    Intended to be consistent with the documentation.
+    https://jira.mesosphere.com/browse/DCOS-51647
+    """
+    backup_name = backup_local_path.name
+    backup_remote_path = Path('/etc/') / backup_name
+
+    for master in all_masters:
+        master.send_file(
+            local_path=backup_local_path,
+            remote_path=backup_remote_path,
+        )
+
+    for master in all_masters:
+        master.run(args=['systemctl', 'stop', 'dcos-exhibitor'])
+
+    for master in all_masters:
+        master.run(
+            args=[
+                '/opt/mesosphere/bin/dcos-shell',
+                'dcos-zk', 'restore', str(backup_remote_path), '-v',
+            ],
+            output=Output.LOG_AND_CAPTURE,
+        )
+
+    for master in all_masters:
+        master.run(args=['systemctl', 'start', 'dcos-exhibitor'])
+
+
+class TestZooKeeperBackup:
+    """
+    Within the context of DC/OS ZooKeeper can be backed up on a running
+    cluster and a previous state can be restored.
+    """
+
+    def test_transaction_log_backup_and_restore(
+        self,
+        artifact_path: Path,
+        docker_backend: Docker,
+        tmp_path: Path,
+        request: SubRequest,
+        log_dir: Path,
+    ) -> None:
+        """
+        In a 3-master cluster, backing up the transaction log of ZooKeeper on
+        one node and restoring from the backup on all master results in a
+        functioning DC/OS cluster with previously backed up Znodes restored.
+        """
+        with Cluster(
+            cluster_backend=docker_backend,
+            masters=3,
+            agents=0,
+            public_agents=0,
+        ) as cluster:
+            cluster.install_dcos_from_path(
+                dcos_installer=artifact_path,
+                dcos_config=cluster.base_config,
+                ip_detect_path=docker_backend.ip_detect_path,
+            )
+            wait_for_dcos_oss(
+                cluster=cluster,
+                request=request,
+                log_dir=log_dir,
+            )
+            master = next(iter(cluster.masters))
+            master_ips = [str(m.public_ip_address) for m in cluster.masters]
+            zk_hostports = ','.join([m + ':2181' for m in master_ips])
+
+            # Write to ZooKeeper before backup
+            persistent_flag = _zk_set_flag(zk_hostports)
+            ephemeral_flag = _zk_set_flag(zk_hostports, ephemeral=True)
```

Comment:
You can remove `ephemeral_flag` test.

> ZooKeeper also has the notion of ephemeral nodes. These znodes exists as long as the session that created the znode is active. When the session ends the znode is deleted. Ephemeral nodes are useful when you want to implement [tbd].

With the current implementation of `_zk_set_flag` the ephemeral node is removed from ZK at the moment when you stop zk client.

Reply:
Unfortunately the ephemeral node is in the backup but cannot be restored because of course we need to restart all ZooKeeper instances. It was not clear to me what would happen to a client in that case. The test was mere a confirmation that ephemeral nodes are not preserved and I would like to keep it to make the behavior clear.